crates/ark/src/lsp/completions/sources/composite.rs

```diff
@@ -14,25 +14,48 @@ mod snippets;
 mod subset;
 mod workspace;
 
-use std::collections::HashSet;
+use std::collections::HashMap;
 
 use stdext::*;
 use tower_lsp::lsp_types::CompletionItem;
 use tower_lsp::lsp_types::CompletionItemKind;
 use tree_sitter::Node;
 
 use crate::lsp::completions::completion_context::CompletionContext;
-use crate::lsp::completions::sources::push_completions;
+use crate::lsp::completions::sources::collect_completions;
+use crate::lsp::completions::sources::CompletionSource;
 use crate::treesitter::NodeType;
 use crate::treesitter::NodeTypeExt;
 
+// Locally useful data structures for tracking completions and their source(s)
+#[derive(Hash, Eq, PartialEq, Clone)]
+struct CompletionKey {
+    label: String,
+    insert_text: Option<String>,
+}
```

Comment:
Previously the key was just `label`. Now its `label` and the `insert_text`. I wonder if this can cause duplication. In theory you can have the same `label` with two different `insert_text` and wed send both to the frontend, and i doubt it deduplicates those on that side. That doesn't feel quite right to me